python/dgl/data/qm9_edge.py

```diff
@@ -0,0 +1,404 @@
+                target = F.tensor(target, dtype=F.data_type_dict['float32'])
+                target = F.cat([target[:, 3:], target[:, :3]], dim=-1)
+                target = (target * conversion.view(1, -1)).tolist()
+
+            with open(os.path.join(self.raw_dir, "uncharacterized.txt"), 'r') as f:
+                skip = [int(x.split()[0]) - 1 for x in f.read().split('\n')[9:-2]]
+
+            suppl = Chem.SDMolSupplier(os.path.join(self.raw_dir, "gdb9.sdf"), removeHs=False, sanitize=False)    
+        
+            N_node = []
+            N_edge = []
+            Node_pos = []
+            Node_attr = []
+
+            src = []
+            dst = []
+            
+            Edge_attr = []
+            Target = []
+            
+            print('Loading graphs:')
+            for i, mol in enumerate(suppl):
+                if i in skip:
+                    continue
+                    
+                N = mol.GetNumAtoms()
+
+                pos = suppl.GetItemText(i).split('\n')[4:4 + N]
+                pos = [[float(x) for x in line.split()[:3]] for line in pos]
+
+                type_idx = []
+                atomic_number = []
+                aromatic = []
+                sp = []
+                sp2 = []
+                sp3 = []
+                num_hs = []
+                
+                for atom in mol.GetAtoms():
+                    type_idx.append(self.types[atom.GetSymbol()])
+                    atomic_number.append(atom.GetAtomicNum())
+                    aromatic.append(1 if atom.GetIsAromatic() else 0)
+                    hybridization = atom.GetHybridization()
+                    sp.append(1 if hybridization == HybridizationType.SP else 0)
+                    sp2.append(1 if hybridization == HybridizationType.SP2 else 0)
+                    sp3.append(1 if hybridization == HybridizationType.SP3 else 0)
+
+                z = F.tensor(atomic_number, dtype=F.data_type_dict['int64'])
+
+                row, col, edge_type = [], [], []
+                for bond in mol.GetBonds():
+                    start, end = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+                    row += [start, end]
+                    col += [end, start]
+                    edge_type += 2 * [self.bonds[bond.GetBondType()]]
+
+                edge_index = F.tensor([row, col], dtype=F.data_type_dict['int64'])
+                edge_type = F.tensor(edge_type, dtype=F.data_type_dict['int64'])
+                edge_attr = np.eye(len(self.bonds))[edge_type]
+
+                perm = (edge_index[0] * N + edge_index[1]).argsort()
+                edge_index = edge_index[:, perm]
+                edge_attr = edge_attr[perm]
+                
+                row, col = edge_index
+                
+                hs = (z == 1).to(F.data_type_dict['float32'])
+                out = F.zeros([int(col.max())+1], dtype=hs[row].dtype, ctx='cpu')
+                num_hs = np.array(out.scatter_add(-1, col, hs[row]))
```

Comment:
I guess `out.scatter_add` will not work for all frameworks. @yzh119 Do we currently support `F.scatter_add`?

Reply:
Yes we do, @hengruizhang98 could you please explain what is `out` and how does `hs`/`num_hs` work?

Reply:
num_hs[i] indicates the number of hydrogen atoms connected to node[i].
I've rewritten it with numpy:

hs = (np.array(atomic_number) == 1).astype(np.int64)

num_hs = np.zeros(n_atom)
for i in range(col.shape[0]):
    num_hs[col[i]] += hs[row][i]

Reply:
So the feature size is one, and we have an API `scatter_add` for that:
https://github.com/dmlc/dgl/blob/master/python/dgl/backend/backend.py#L1555-L1575

Reply:
Thanks! Now it has been changed into: 

x = F.tensor(hs[row], dtype=F.data_type_dict['float32'])
idx = F.tensor(col, dtype=F.data_type_dict['int64'])
num_hs = F.scatter_add(x, idx, n_atom)

Reply:
I'm okay with about change.